automata/base/automaton.py

```diff
@@ -33,11 +33,17 @@ def validate(self):
 
     def __setattr__(self, name, value):
         """Set custom setattr to make class immutable."""
-        raise AttributeError(f'This {type(self).__name__} is immutable')
+        if global_config.allow_mutable_automata:
```

Comment:
I considered calling self.validate() here to alert users if they perform an invalid operation. But I opted not to for 2 reasons:
1- Performance
2- Because loading DFAs with pickle would require to disable validation too (as pickle calls "setattr"). It could be done, but it makes it more cumbersome to use

Also, I go by the assumption that if people disable the immutability of the automata, then they should really know what they are doing

Reply:
@EduardoGoulart1 Agreed. It would hurt performance to call `validate()` on every assignment, especially if the user needs to make multiple assignments in succession. My assumption has always been, for a mutable automaton, that `validate()` only runs under the following conditions:

1. Once in instantiation
2. Manually by the user

That is precisely why `validate()` is not `_validate()`—because it can be called by the user as needed. All that to say, yes, I agree with your assessment.

Reply:
@caleb531 I decided against having set/del attributes because the automata such as DFA class have the word cache which we don't want to mess-up with. Instead, to enable pickle, I directly use the __getstate__ and __setstate__ methods.